controller/health.go

```diff
@@ -0,0 +1,64 @@
+package controller
+
+import (
+	"github.com/argoproj/gitops-engine/pkg/health"
+	hookutil "github.com/argoproj/gitops-engine/pkg/sync/hook"
+	"github.com/argoproj/gitops-engine/pkg/sync/ignore"
+	kubeutil "github.com/argoproj/gitops-engine/pkg/utils/kube"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/argoproj/argo-cd/v2/pkg/apis/application"
+	appv1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/v2/util/lua"
+)
+
+// setApplicationHealth updates the health statuses of all resources performed in the comparison
+func setApplicationHealth(resources []managedResource, statuses []appv1.ResourceStatus, resourceOverrides map[string]appv1.ResourceOverride, app *appv1.Application) (*appv1.HealthStatus, error) {
+	var savedErr error
+	appHealth := appv1.HealthStatus{Status: health.HealthStatusHealthy}
+	for i, res := range resources {
+		if res.Target != nil && hookutil.Skip(res.Target) {
+			continue
+		}
+
+		if res.Live != nil && (hookutil.IsHook(res.Live) || ignore.Ignore(res.Live)) {
+			continue
+		}
+
+		var healthStatus *health.HealthStatus
+		var err error
+		healthOverrides := lua.ResourceHealthOverrides(resourceOverrides)
+		gvk := schema.GroupVersionKind{Group: res.Group, Version: res.Version, Kind: res.Kind}
+		if res.Live == nil {
+			healthStatus = &health.HealthStatus{Status: health.HealthStatusMissing}
+		} else {
+			// App the manages itself should not affect own health
+			if isSelfReferencedApp(app, kubeutil.GetObjectRef(res.Live)) {
+				continue
+			}
+			healthStatus, err = health.GetResourceHealth(res.Live, healthOverrides)
+			if err != nil && savedErr == nil {
+				savedErr = err
+			}
+		}
+		if healthStatus != nil {
+			resHealth := appv1.HealthStatus{Status: healthStatus.Status, Message: healthStatus.Message}
+			statuses[i].Health = &resHealth
+
+			// Is health status is missing but resource has not built-in/custom health check then it should not affect parent app health
+			if _, hasOverride := healthOverrides[lua.GetConfigMapKey(gvk)]; healthStatus.Status == health.HealthStatusMissing && !hasOverride && health.GetHealthCheckFunc(gvk) == nil {
+				continue
+			}
+
+			// Missing or Unknown health status of child Argo CD app should not affect parent
```

Comment:
Hello, I believe this block breaks the use of sync waves with the app-of-apps pattern. Specifically during initial creation of multiple layers of Applications (e.g. "app-of-apps-of-apps"). If sync waves are used on an Application in the intermediate layer of the hierarchy, the initial `Missing` / `Unknown` health statuses of its child Applications don't bubble up, and the sync proceeds to the other intermediate layer Applications even if children Applications haven't been synced and become healthy.

Why does this block need to exist?  Adding back health checks for the app-of-apps pattern is already optional (https://argo-cd.readthedocs.io/en/stable/operator-manual/health/#argocd-app). If a user hasn't added them back, it seems the preceding block on lines 48-51 would already exclude child Application health statuses from affecting the parent's. So, unless I'm missing something, this block on lines 54-56 only has an effect when the user has opted into the app-of-apps health checks, but that effect is to partially break sync waves on Application resources, which is pretty vital for the app-of-apps pattern.